src/core/cgroup.c

```diff
@@ -1172,17 +1173,42 @@ static void cgroup_context_apply(
                 cgroup_apply_firewall(u);
 }
 
+static bool need_cpu_controller_for_accounting_cached(void) {
+        static int needed = -1;
+        if (needed < 0)
+                needed = need_cpu_controller_for_accounting();
+        return needed;
+}
+
+bool need_cpu_controller_for_accounting(void) {
+        /* On 4.15+ with unified hierarchy, cpu.stat's usage_usec is provided
+         * externally from the CPU controller, which means we don't need to
+         * enable the CPU controller just to get metrics.
+         */
+        struct utsname u;
+        int kernel_too_old;
+
+        assert_se(uname(&u) >= 0);
+        kernel_too_old = str_verscmp(u.release, "4.15") < 0;
+        return kernel_too_old || !cg_all_unified();
+}
```

Comment:
why split this function in two? i.e. in the uncached and the cached version

Reply:
`kernel_too_old` could be `bool` actually...

That said, you could also just get rid of it entirely, and just write: 

```c
return str_verscmp(u.release, "4.15") < 0 || !cg_all_unified();
```

Reply:
btw, thinking about this more, I think it would be really nice if this function would be changed to return a `CGroupMask` that would indicate the precise mask required to get cpu accounting. It would then return:

* in cgroupsv1 mode `CGROUP_MASK_CPUACCT`
* in cgroupsv2 mode on old kernels `CGROUP_MASK_CPU`
* in cgroupsv2 mode on new kernels `0`

I think this would neatly simplify stuff, as we can generalize the realization check in `unit_get_cpu_usage_raw()` with that, as well as the invalidation in bus_cgroup_set_property()...

Sorry for the forth and back, I only realized that now...

Reply:
Oh yeah, that's a good idea. I already didn't like that we duplicate the logic in dbus-cgroup.c.